core/templates/dev/head/domain/topic/TopicUpdateService.js

```diff
@@ -230,6 +228,61 @@ oppia.factory('TopicUpdateService', [
           // Get the current change list.
           var currentChangeList = UndoRedoService.getChangeList();
           var indicesToDelete = [];
+          var oldSubtopicIdSkillIdMap = {};
+          var index = null;
+          var skillIds = subtopic.getSkillIds();
+          for (var i = 0; i < currentChangeList.length; i++) {
+            var changeDict =
+              currentChangeList[i].getBackendChangeObject();
+            if (oldSubtopicIdSkillIdMap[changeDict.skill_id] === undefined) {
+              if (changeDict.new_subtopic_id === subtopicId) {
+                // Record the origin subtopic id for the skill.
+                oldSubtopicIdSkillIdMap[changeDict.skill_id] =
+                  changeDict.old_subtopic_id;
+                index = i;
+              }
+            } else {
+              if (changeDict.cmd === CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC) {
+                // If the skill was removed from all subtopics at some point,
+                // modify that change dict, so that it was as if it was removed
+                // from its recorded origin.
+                changeDict.subtopic_id =
+                  oldSubtopicIdSkillIdMap[changeDict.skill_id];
+              } else {
+                // At every move operation, set old subtopic id as the recorded
```

Comment:
Shouldn't there be filtering to ensure that this is actually a move operation?

Reply:
Similar to my previous reply, only if a new_subtopic_id field was encountered, would oldSkillIdToSubtopicIdMapping have that skill id field, and new_subtopic_id is only present in a move command. Should I add a check for cmd anyways? though I don't think new_subtopic_id would be used anywhere else.

Reply:
Done, added checks for commands.